Title: Profitable App Profiles for the App Store and Google Play Markets

Using basic Python, List and for loops, Conditional statement, Dictionaries, and Functions

We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. 

This means our revenue for any given app is mostly influenced by the number of users who use our app — the more users that see and engage with the ads, the better. 

Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.

-------------------------------------------------------------------------------
Data Exploration

Opening two datasets, App Store dataset and Google Play dataset, saving them as lits of lists, and exploring them

In [67]:
from csv import reader

opening_as = open('AppleStore.csv', encoding='utf8')
reading_as = reader(opening_as)
aps = list(reading_as)
print(aps[0])
print('\n')
print(aps[1:5])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1'], ['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']]


In [68]:
opening_gp = open('googleplaystore.csv', encoding='utf8')
reading_gp = reader(opening_gp)
gp = list(reading_gp)
print(gp[0])
print('\n')
print(gp[1:5])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]


-------------------------------------------------------------------------------
Data Cleaning

Detect inaccurate data to either correct it or remove it.

Detect duplicate data and remove the duplicate.

Remove non-English apps.

Remove non-free apps

-------------------------------------------------------------------------------
Deleting Inaccurate row.

Row Index Number 10473 has an error.

Deleting Index Number 10473.

In [69]:
print(len(gp[10472]))
print(len(gp[10473]))

13
12


In [70]:
gp[10472]

['Xposed Wi-Fi-Pwd',
 'PERSONALIZATION',
 '3.5',
 '1042',
 '404k',
 '100,000+',
 'Free',
 '0',
 'Everyone',
 'Personalization',
 'August 5, 2014',
 '3.0.0',
 '4.0.3 and up']

In [71]:
# Second row is missing which is supposed to be the app's category
gp[10473]

['Life Made WI-Fi Touchscreen Photo Frame',
 '1.9',
 '19',
 '3.0M',
 '1,000+',
 'Free',
 '0',
 'Everyone',
 '',
 'February 11, 2018',
 '1.0.19',
 '4.0 and up']

In [72]:
del gp[10473]

In [73]:
gp[10473]

['osmino Wi-Fi: free WiFi',
 'TOOLS',
 '4.2',
 '134203',
 '4.1M',
 '10,000,000+',
 'Free',
 '0',
 'Everyone',
 'Tools',
 'August 7, 2018',
 '6.06.14',
 '4.4 and up']

-------------------------------------------------------------------------------
Removing Duplicates based on the number of reviews.

Keeping the one that has the highest number of reviews.

Counting the numbers of unique apps and duplicate apps.

In [74]:
duplicate_row = []
unique_row = []
unique_app = []

for i in gp[1:]:
    if i[0] in unique_app:
        duplicate_row.append(i)
    else:
        unique_app.append(i[0])
print('The number of unique google play store apps is', len(unique_app))
print('The number of duplicate google play store apps is', len(duplicate_row))

The number of unique google play store apps is 9659
The number of duplicate google play store apps is 1181


In [75]:
duplicate_row2 = []
unique_row2 = []
unique_app2 = []

for i in aps[1:]:
    if i[0] in unique_app2:
        duplicate_row2.append(i)
    else:
        unique_app2.append(i[0])
print('The number of unique apple store apps is', len(unique_app2))
print('The number of duplicate apple store apps is', len(duplicate_row2))

The number of unique apple store apps is 7197
The number of duplicate apple store apps is 0


Creating a dictionary where each key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.

Expected number of apps in reviews_max dictionary should be 9659.

In [76]:
reviews_max = {}

for i in gp[1:]:
    name = i[0]
    n_reviews = float(i[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


Using dictionary above to remove duplicate rows.

Again, expected length is 9659.

In [77]:
android_clean = []
already_added = []

for i in gp[1:]:
    name = i[0]
    n_reviews = float(i[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(i)
        already_added.append(name)
print(len(android_clean))
print(len(already_added))

9659
9659


-------------------------------------------------------------------------------
Removing Non-English apps.

The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. 

If an app name contains a character that is greater than 127, then it probably means that the app has a non-English name.

Writing a function that takes in a string and returns False if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns True.

In [78]:
def detector(string):
    count = 0
    for i in string:
        if ord(i) > 127:
            count += 1
        else:
            continue
    if count > 3:
        return False
    else:
        return True

instagram = detector('Instagram')
print(instagram)

non_eng = detector('爱奇艺PPS -《欢乐颂2》电视剧热播')
print(non_eng)

True
False


Using the function above to weed out non-English apps from both data sets.

In [79]:
android_clean2 = []

for i in android_clean:
    if detector(i[0]) == True:
        android_clean2.append(i)
print(len(android_clean2))

9614


In [80]:
aps_clean = []

for i in aps:
    if detector(i[1]) == True:
        aps_clean.append(i)
print(len(aps_clean))

6184


-------------------------------------------------------------------------------
Removing non-Free apps.

In [81]:
android_final = []
aps_final = []

for i in android_clean2:
    price = i[7]
    if price == '0':
        android_final.append(i)

for j in aps_clean:
    price = j[4]
    if price == '0.0':
        aps_final.append(j)
        
print(len(android_final))
print(len(aps_final))

8864
3222


-------------------------------------------------------------------------------
Data Analysis

What are the most common genres for each market.

Generating frequency tables.

In [82]:
android = {}
apple = {}

for i in android_final:
    if i[1] not in android:
        android[i[1]] = 1
    elif i[1] in android:
        android[i[1]] += 1
        
for j in aps_final:
    if j[11] not in apple:
        apple[j[11]] = 1
    elif j[11] in apple:
        apple[j[11]] += 1

android

{'ART_AND_DESIGN': 57,
 'AUTO_AND_VEHICLES': 82,
 'BEAUTY': 53,
 'BOOKS_AND_REFERENCE': 190,
 'BUSINESS': 407,
 'COMICS': 55,
 'COMMUNICATION': 287,
 'DATING': 165,
 'EDUCATION': 103,
 'ENTERTAINMENT': 85,
 'EVENTS': 63,
 'FINANCE': 328,
 'FOOD_AND_DRINK': 110,
 'HEALTH_AND_FITNESS': 273,
 'HOUSE_AND_HOME': 73,
 'LIBRARIES_AND_DEMO': 83,
 'LIFESTYLE': 346,
 'GAME': 862,
 'FAMILY': 1676,
 'MEDICAL': 313,
 'SOCIAL': 236,
 'SHOPPING': 199,
 'PHOTOGRAPHY': 261,
 'SPORTS': 301,
 'TRAVEL_AND_LOCAL': 207,
 'TOOLS': 750,
 'PERSONALIZATION': 294,
 'PRODUCTIVITY': 345,
 'PARENTING': 58,
 'WEATHER': 71,
 'VIDEO_PLAYERS': 159,
 'NEWS_AND_MAGAZINES': 248,
 'MAPS_AND_NAVIGATION': 124}

In [83]:
apple

{'Social Networking': 106,
 'Photo & Video': 160,
 'Games': 1874,
 'Music': 66,
 'Reference': 18,
 'Health & Fitness': 65,
 'Weather': 28,
 'Utilities': 81,
 'Travel': 40,
 'Shopping': 84,
 'News': 43,
 'Navigation': 6,
 'Lifestyle': 51,
 'Entertainment': 254,
 'Food & Drink': 26,
 'Sports': 69,
 'Book': 14,
 'Finance': 36,
 'Education': 118,
 'Productivity': 56,
 'Business': 17,
 'Catalogs': 4,
 'Medical': 6}

Using functions to create frequency tables based on Prime Genre, Genres, and Category

In [84]:
def freq_table(dataset, index):
    dictionary = {}
    count = 0
    
    for i in dataset:
        if i[index] in dictionary:
            dictionary[i[index]] += 1
            count += 1
        else:
            dictionary[i[index]] = 1
            count += 1
    
    for j in dictionary:
        dictionary[j] = (dictionary[j] / count) * 100
    return dictionary

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

android_category = display_table(android_final, 1)
android_category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [85]:
android_genres = display_table(android_final, -4)
android_genres

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In [88]:
apple_prime_genre = display_table(aps_final, -5)
apple_prime_genre

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Most common genre in App Store dataset: Games

Runner up: Entertainment

Pattern: Game - 58%, Entertainment - 7.8%, Photo & Video - 4.9%, Social Networking - 3.2%

Genral Impression: Most of the apps people use in App Store are for fun and entertainment

Most common genre in Google Play dataset: Family and Tool

Runner up: Game and Entertainment

Pattern: Family - 18%, Tools - 8.5%, Entertainment - 6%, Education - 5.3%

Genral Impression: Not many are for fun. They're mostly practical purposes.

-------------------------------------------------------------------------------
What kind of apps with most users?

For the Google Play data set, we can find this information in the Installs column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_column app.

We will calculate the average number of user ratings per app genre on the App Store:

Isolate the apps of each genre, Sum up the user ratings for the apps of that genre, and Divide the sum by the number of apps belonging to that genre (not by the total number of apps).

In [99]:
freq_table_dict = {}

for i in aps_final:
    if i[-5] in freq_table_dict:
        freq_table_dict[i[-5]] += 1
    else:
        freq_table_dict[i[-5]] = 1
        
for genre in freq_table_dict:
    total = 0
    len_genre = 0
    for i in aps_final:
        genre_app = i[-5]
        if genre_app == genre:
            nour = float(i[5])
            total += nour
            len_genre += 1
    avg_num_user_rating = total/len_genre
    print(genre, ":", avg_num_user_rating)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


We have data about the number of installs for the Google Play market, so we should be able to get a clearer picture about genre popularity. 

However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.)

In [114]:
ft_category = {}
for i in android_final:
    if i[1] in ft_category:
        ft_category[i[1]] += 1
    else:
        ft_category[i[1]] = 1

for category in ft_category:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_